docs: Improve documentation for writing and running tests

andrewleech opened the pull request:
### Summary

I wanted to use some LLM assitance to create new unit tests. To aid in this process, it helps to have comprehensive docs to inform the LLM's how best to write and run unit tests, so I started with that.

These updates were mostly written with Google's gemini-2.5-pro-exp-03-25 though some final review and cleanup was done with Anthropic Claude 3.7

The updated documentation provides:
- More detailed explanation of the test directory structure and key runner scripts.
- Clearer instructions on writing simple tests, including the use of print() and .exp files.
- Expanded coverage of run-tests.py options for targeting specific devices, filtering tests, and handling failures.
- An overview of how run-tests.py works internally (feature detection, skipping, execution, comparison).
- Guidance on writing advanced tests (skipping, unittest, special handling, float/endian considerations).
- Introduction to multi-instance testing with run-multitests.py, including how to write tests with instanceN() functions and use the `multitest` helper for coordination.
- Examples of running multi-instance tests with different instance types.

### Testing

Just manual review of the content.

agatti commented:
Not sure if it's important enough to be featured in that document, but maybe it'd be useful to document the special test file name prefixes/suffixes recognised by `tests/run-tests.py` that act as rudimentary feature checks to skip certain tests?

For example, test files whose name starts with `native_` or `viper_` will be automatically skipped if `@micropython.native` or `@micropython.viper` support was not detected in the target interpreter, and so on.  There's a dozen or so of those tags so maybe it's not worth documenting them all, but again, the LLM may sort all that stuff out by itself :)

andrewleech commented:
> special test file name prefixes/suffixes

Yeah I do think it'd be helpful to have these itemised somewhere as it's certainly not abvious for new developers! 

I'm not too sure about how best to divide what does in the rst docs and what goes in the readme here... I kinda think the readme is perhaps better for "internals" things like that which are maybe of more interest to low level developers, with the rst docs more focussing on "end users" interested in running tests?

agatti commented:
I personally use the README files as a quickstart guide myself, always favouring explicit documentation if it's available.  It used to be like that back then, when dinosaurs still roamed freely and whatnot, so maybe somebody from a younger demographics should weigh in on this.

Considering the users/developers split, I wonder if this possible TOC for three (or two) documents - names are generic on purpose - can be a starting point for a discussion on this:

* "Quickstart"/"End users"
	- Standard tests
		- How to run tests (simplified version, no need to bring emitters and mpy in, f.ex.)
		- How a successful run should look like
		- Where to look for failure info to send back to a developer
	- Performance benchmarks
		- How to run benchmarks (simplified version)
		- How to read output and compare with previous runs
* "Developers"
	- How to run tests (what's left to document)
	- Standard tests (almost everything that's in `writingtests.rst` for standard tests, maybe skip more niche sub-topics)
	- Multi-instance tests
	- SSL tests
	- Performance benchmarks (what's left to document)
* "Internals" (fold this into "Developers" if needed)
	- How `run-tests.py` enumerates tests and how it decides which ones to skip
	- Natmod tests
	- Internal benchmarks
	- Everything else that's not in "Developers" ?

I can help out on the natmod tests bit if needed, as long as somebody else fixes the grammar :)

Josverl commented:
for context : I just got an inflatable Dino for my birthday.
There is no need to settle on just a single topilogy 
With Sphinx & MyST it is simple to pull in the Readme.MD and other markdown docs into the online documentation.  
That allows the information  to be found and accessed va different ways and personal preferences , without increasing the effort to maintain it.
( possibly that is also a way to pull in some of the information in the Wiki , which is rather hard to search )

agatti commented:
FYI, maybe you may want to update the documentation to also cover the extra argument introduced in 6fd7422a8e5b6c8952f7b67713f0551992d247cc.  This is probably something that should be put into the "advanced" section :)

projectgus commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,355 @@
+   - Endianness
+
+2. **Automatic Skipping**: Tests are skipped based on:
+   
```
Suggest documenting the explicit SKIP output first, as that's the preferred method. Something like:

---

Where possible, a test should determine at runtime if it should be skipped. If so, it prints a single line `SKIP` before existing with `raise SystemExit()`. The test runner parses this output and skips the test.

This isn't always possible, particularly if the test relies on optional MicroPython features that prevent parsing the .py file (i.e. importing the file raises SyntaxError before the skip logic can run.) In these cases, there is additional logic for skipping tests:

andrewleech replied:
Reordered, the runtime SKIP is documented first now with the filename patterns as the fallback for when the file cant be parsed. Used your wording pretty much as-is, thanks.

projectgus commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,355 @@
+5. **Documentation**: Comment complex test logic
+6. **Cross-platform**: Test on multiple architectures when possible
+
+Debugging Test Failures
```
Maybe this section could be merged into test_quickstart.rst, as it's relevant for people running tests as well?

andrewleech replied:
Agreed, that section is in test_quickstart.rst now.

projectgus reviewed (commented):
Hi @andrewleech,

This is really good, and I think overdue! (I noted that it's inspired by improving the docs for bots not for humans, but either way it's good to have better docs!)

There's some newer features and changes that would need to be folded back in here (like `target_wiring.py` and I think some of the skip checks have been simplified out since this was written), but it's looking very solid.

It's taken a while for me to notice it (sorry), are you still interested in updating it?

pi-anl committed:
docs/develop/writingtests: Rework test documentation.

Split the test documentation into a guide for running the tests
(test_quickstart) and one for writing and maintaining them
(test_developers), with writingtests.rst reduced to a toctree over the
two. Both note up front that they cover MicroPython's own test suite,
not testing of applications that run on MicroPython.

Between them they document the three ways a test determines pass/fail
(CPython output comparison, unittest summary, or a .exp file), runtime
SKIP with the filename patterns as a fallback, run-tests.py options and
target selection, multi-instance tests, target_wiring, native module
tests and the internal benchmarks.

tests/README.md is trimmed to orientation and running the tests, and
references the full docs for the rest.

Signed-off-by: Andrew Leech <andrew.leech@planet-innovation.com>
Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech commented:
Thanks for the feedback everyone, I've reworked the docs in the latest push.

@projectgus yes definitely still interested in this, sorry for the delay.

Main changes:
- Restructured into quickstart (running tests) and developer guide (writing tests, advanced usage), with writingtests.rst as a landing page. The split follows @agatti's suggested topology.
- Improved accuracy of CLI examples and test file references throughout, verified against the actual test runner source.
- Added the filename-based skip patterns list per @agatti's suggestion.
- Added `--begin` injected code customisation per @agatti's note about 6fd7422a8e.
- Added `output_metric` to the multitest API listing.
- Tightened the prose throughout, removed generic filler sections.
- README kept as a lean GitHub-browsable quickstart with a link to the full RST docs.

I haven't added rst include for the README as @Josverl suggested since it would need myst_parser added as a sphinx extension, and the relative links in the README would break when included. The duplication between README and quickstart is minimal now and intentional for the two different audiences.

projectgus reviewed (commented):
This is really great @andrewleech, thanks again for submitting it!

I have a bunch more comments but they're all quite minor, I'd also be very happy merging this in its current form if you don't have time to follow up.

andrewleech commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,387 @@
+
+   print("Testing feature X")
+   result = some_function()
+   print(result)
```
Good idea, there is a full example now: micropython/heap_locked.py with the contents of its .py.exp alongside it.

andrewleech commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,387 @@
+- ``int_64*``, ``*_int64`` - require 64-bit integer support
+- ``*reverse_op*`` - require ``__rOP__`` special methods
+- ``string_fstring*`` - require f-string support
+- ``asm*`` - require inline assembly for the target architecture
```
Hehe, converted to a grid table. Did the target selection and multitest API lists the same way while I was in there.

andrewleech commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,387 @@
+
+The ``multitest`` module provides:
+
+- ``next()`` - synchronize instances at a stage boundary
```
Applied, thanks.

andrewleech commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,387 @@
+- ``get_network_ip()`` - get the instance's IP address
+- ``expect_reboot(resume, delay_ms=0)`` - handle device reboots during test
+- ``output_metric(data)`` - emit a metric string for collection by the runner
+- ``skip()`` - skip the test from any instance
```
Applied.

andrewleech commented on docs/develop/test_quickstart.rst:
```diff
@@ -0,0 +1,150 @@
+
+Running Tests
+=============
+
```
Added, the intro now says these are the tests for MicroPython itself rather than for your own application, and why you would run them.

andrewleech commented on docs/develop/test_developers.rst:
```diff
@@ -0,0 +1,387 @@
+Writing and Maintaining Tests
+=============================
+
+This covers writing tests and advanced test runner usage. For running
```
Same here, there is a scope note at the top now which refers over to test_quickstart for actually running them.

andrewleech commented on tests/README.md:
```diff
@@ -1,251 +1,134 @@
-# MicroPython Test Suite
+# MicroPython Test Suite - Quick Start Guide
```
Applied.

andrewleech commented:
Rebased onto current master.

Note that 641fa01c0b added a paragraph to `writingtests.rst` about covering non-ASCII / Unicode text in string-related tests, which landed after this branch was originally based. As this PR reduces `writingtests.rst` to a toctree, I have carried that guidance across into `test_developers.rst` as a short Unicode section rather than dropping it.

Josverl commented:
don't want to cause scope creep but I miss 2 test topics.
- testing mpremote
- testing modules in micropython-lib

these can/should be their own pages in the tree - but I wonder if the current page names are clear enough fur a future addition

- Testing MicroPython
  - Running tests
  - Writing and Maintaining Tests
  - _Running `mpremote` tests_
  - _Testing micropython-lib modules_
  - _Writing and Maintaining Tests for micropython-lib modules_
